File: src/types/youtube.ts
// YouTube AI视频制作工具类型定义
// 与FastAPI后端接口对齐

// ============ 枚举类型 ============

// 项目状态枚举（与后端ProjectStatus对齐）
export type ProjectStatus =
  | 'created' // 刚创建，storyboards为空
  | 'prompts_ready' // 提示词已生成，storyboards已初始化
  | 'images_partial' // 部分分镜有图片
  | 'images_ready' // 所有分镜都有图片
  | 'videos_partial' // 部分分镜有视频
  | 'completed' // 所有分镜都有视频
  | 'failed'; // 失败

// 图片生成类型（与后端GenerationType对齐）
export type GenerationType = 'text_to_image' | 'image_text_to_image';

// 主体类型（V2新增）
export type SubjectType = 'character' | 'object' | 'scene';

// 图片比例（V2新增）
export type AspectRatio = '9:16' | '16:9';

// 分镜插入类型（V2新增）
export type InsertType = 'before' | 'after';

// ============ JSONB嵌套数据结构（与后端对齐）============

// 生成的图片
export interface GeneratedImage {
  url: string;
  generation_type: GenerationType;
}

// 生成的视频
export interface GeneratedVideo {
  url: string;
  source_image_index: number; // 记录使用的是哪张图片生成的
}

// 分镜数据（与后端Storyboard对齐）
export interface Storyboard {
  index: number;
  storyboard_summary?: string; // 分镜概述（简短描述该分镜的内容和目的）
  text_to_image: string; // 文生图提示词（详细的图片生成指令）
  image_to_video: string; // 图生视频提示词（详细的视频生成指令）
  character_refs?: string[] | null; // 角色引用标识
  ref_storyboard_indexes?: number[] | null; // 参考分镜索引列表，用于场景一致性
  is_prompt_edited: boolean; // 是否被手动编辑过
  images: GeneratedImage[]; // 生成的图片数组
  selected_image_index: number | null; // 选中的图片索引
  videos: GeneratedVideo[]; // 生成的视频数组
  selected_video_index: number | null; // 选中的视频索引
}

// 项目JSONB数据（与后端ProjectData对齐）
export interface ProjectData {
  name: string;
  youtube_url: string;
  status: ProjectStatus;
  prompt_version?: string | null;
  aspect_ratio?: AspectRatio; // V2新增：图片生成比例
  current_prompt_version?: string; // V2新增：当前提示词版本
  subject_mappings?: Record<string, string>; // V2新增：主体映射 {"角色A": "uuid", ...}
  prompt_history?: PromptHistoryVersion[]; // V2新增：提示词历史版本
  storyboards: Storyboard[];
}

// ============ API响应模型（与后端对齐）============

// 项目响应（与后端ProjectResponse对齐）
export interface ProjectResponse {
  id: string;
  data: ProjectData;
  created_at: string;
  updated_at: string;
}

// 项目列表项（用于列表页面显示）
export interface ProjectListItem {
  id: string;
  name: string;
  youtube_url: string;
  status: ProjectStatus;
  storyboard_count: number;
  cover_url?: string | null; // 封面图片URL（从viral_context.image_host_cover_url提取）
  created_at: string;
  updated_at: string;
}

// 分页响应（与后端PaginatedResponse对齐）
export interface PaginatedProjectResponse {
  items: ProjectListItem[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// 生成提示词响应（与后端GeneratePromptsResponse对齐）
export interface GeneratePromptsResponse {
  success: boolean;
  storyboard_count: number;
  prompt_version?: string;
  error?: string;
}

// 生成图片响应（与后端GenerateImageResponse对齐）
export interface GenerateImageResponse {
  success: boolean;
  storyboard_index: number;
  image?: GeneratedImage;
  image_index?: number;
  error?: string;
}

// 生成视频响应（与后端GenerateVideoResponse对齐）
export interface GenerateVideoResponse {
  success: boolean;
  storyboard_index: number;
  video?: GeneratedVideo;
  video_index?: number;
  error?: string;
}

// ============ 请求模型（与后端对齐）============

// 创建项目请求
export interface CreateProjectRequest {
  name: string;
  youtube_url: string;
}

// 更新项目请求
export interface UpdateProjectRequest {
  name?: string;
  storyboards?: Partial<Storyboard>[];
  aspect_ratio?: AspectRatio; // V2新增
  subject_mappings?: Record<string, string>; // V2新增
}

// 生成提示词请求
export interface GeneratePromptsRequest {
  instruction?: string;
  system_prompt?: string;
}

// 角色引用
export interface CharacterRef {
  identifier: string;
  image_url: string;
}

// 生成图片请求
export interface GenerateImageRequest {
  storyboard_index: number;
  character_images?: string[]; // 角色参考图片数组（支持base64或URL）
  aspect_ratio?: AspectRatio; // V2新增：图片比例
  // 注意：ref_storyboard_indexes 已移至 Storyboard 模型，由后端自动读取
}

// 生成视频请求
export interface GenerateVideoRequest {
  storyboard_index: number;
}

// ============ 兼容旧代码的类型别名 ============

// VideoProject 作为 ProjectResponse 的别名，方便旧代码迁移
export type VideoProject = ProjectResponse;

// 通用分页响应（兼容旧代码）
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// ============ 前端专用类型 ============

// 角色映射（前端Settings页面使用）
export interface CharacterMapping {
  id: string;
  identifier: string; // A, B, C, D 等标识
  reference_image_url?: string; // 角色参考图片URL
  name?: string; // 角色名称（可选）
}

// 工作流步骤状态
export type WorkflowStepStatus =
  | 'completed'
  | 'in_progress'
  | 'pending'
  | 'failed';

// 工作流步骤
export interface WorkflowStep {
  id: string;
  name: string;
  status: WorkflowStepStatus;
  description?: string;
  action_url?: string;
  progress?: number;
}

// 结构化提示词JSON（导出用）
export interface StructuredPromptData {
  project_id: string;
  project_name: string;
  total_storyboards: number;
  character_mappings: CharacterMapping[];
  prompts: {
    storyboard_index: number;
    text_to_image: string;
    image_to_video: string;
    character_refs: string[];
    character_images: { [key: string]: string };
  }[];
}

// API错误响应
export interface ApiError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

// ============ 媒体管理类型 ============

// 媒体统计信息
export interface MediaCategoryStats {
  total_count: number;
  selected_count: number;
  unselected_count: number;
  total_size: string;
  total_bytes: number;
}

// 项目媒体统计响应
export interface MediaStatsResponse {
  project_id: string;
  images: MediaCategoryStats;
  videos: MediaCategoryStats;
  total_size: string;
  total_bytes: number;
}

// 媒体清理类型
export type MediaCleanupType = 'all' | 'images' | 'videos';

// 媒体清理响应
export interface MediaCleanupResponse {
  success: boolean;
  deleted_images: number;
  deleted_videos: number;
  freed_size: string;
  freed_bytes: number;
  errors: string[];
}

// ============ 旧类型定义（保留兼容性）============

// 源视频分镜（仅供参考展示，不影响后续流程）
export interface SourceStoryboard {
  id: string;
  project_id: string;
  index: number;
  start_time: number;
  end_time: number;
  start_frame_url: string;
  end_frame_url: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

// 提示词（旧版本，保留兼容性）
export interface Prompt {
  id: string;
  project_id: string;
  storyboard_id: string;
  storyboard_index: number;
  storyboard_summary?: string; // 分镜概述
  text_to_image: string;
  image_to_video: string;
  character_refs?: string[];
  version: 'v1' | 'v2';
  is_edited: boolean;
  edit_history?: PromptEditHistory[];
  created_at: string;
  updated_at: string;
}

// 提示词编辑历史
export interface PromptEditHistory {
  timestamp: string;
  text_to_image: string;
  image_to_video: string;
  edit_type: 'manual' | 'ai_regenerate';
}

// 生成任务状态
export type TaskStatus =
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'paused'
  | 'cancelled';

// 生成任务
export interface GenerationTask {
  id: string;
  project_id: string;
  task_type: 'image' | 'video' | 'download' | 'parse' | 'prompt';
  status: TaskStatus;
  progress: number;
  total_items: number;
  completed_items: number;
  failed_items: number;
  error_message?: string;
  created_at: string;
  updated_at: string;
}

// 任务响应
export interface TaskResponse {
  task_id: string;
  status: string;
  message: string;
}

// 下载响应
export interface DownloadResponse {
  download_url: string;
  file_count: number;
  total_size: string;
}

// 更新分镜请求（旧版本）
export interface UpdateStoryboardRequest {
  description?: string;
  start_time?: number;
  end_time?: number;
}

// 重新生成提示词请求
export interface RegeneratePromptRequest {
  instruction: string;
  regenerate_type: 'text_to_image' | 'image_to_video' | 'both';
}

// 更新角色映射请求
export interface UpdateCharacterMappingsRequest {
  mappings: {
    number: number;
    identifier: string;
    name?: string;
  }[];
}

// 批量生成请求
export interface BatchGenerateRequest {
  storyboard_ids?: string[];
  parallel_count?: number;
}

// 视频生成请求（包含源图片和提示词）
export interface VideoGenerateRequestLegacy {
  items: {
    prompt_id: string;
    source_image_id: string;
    image_to_video_prompt: string;
  }[];
  parallel_count?: number;
}

// 结构化提示词中的单个提示词项
export interface StructuredPromptItem {
  storyboard_index: number;
  text_to_image: string;
  image_to_video: string;
  character_refs: string[];
  character_images: { [key: string]: string };
}

// ============ V2 新增类型定义 ============

// ============ 全局主体库类型 ============

// 全局主体（服务端存储）
export interface Subject {
  id: string; // UUID
  type: SubjectType;
  identifier?: string; // 已废弃，保留兼容
  name?: string;
  image_url?: string;
  created_at: string;
  updated_at: string;
}

// 主体列表响应
export interface SubjectListResponse {
  subjects: Subject[];
  total: number;
}

// 创建主体请求（FormData，这里定义字段）
export interface CreateSubjectFields {
  type: SubjectType;
  name?: string;
  // image 通过 FormData 传递
}

// 更新主体请求（FormData，这里定义字段）
export interface UpdateSubjectFields {
  name?: string;
  remove_image?: boolean;
  // image 通过 FormData 传递
}

// ============ 提示词历史版本类型 ============

// 对话消息
export interface ConversationMessage {
  role: 'user' | 'assistant';
  content: string;
}

// 历史版本中的分镜（不含图片/视频）
export interface PromptHistoryStoryboard {
  index: number;
  storyboard_summary?: string; // 分镜概述
  text_to_image: string;
  image_to_video: string;
  character_refs?: string[] | null;
  is_prompt_edited: boolean;
}

// 提示词历史版本
export interface PromptHistoryVersion {
  version: string; // v1, v2, v3...
  created_at: string;
  parent_version?: string; // 父版本号（v1无此字段）
  instruction: string; // 生成时的用户指令
  conversation_history: ConversationMessage[]; // Gemini对话历史
  storyboards: PromptHistoryStoryboard[]; // 该版本的分镜提示词
}

// 历史版本摘要（用于列表展示）
export interface PromptHistorySummary {
  version: string;
  created_at: string;
  instruction: string;
  storyboard_count: number;
  parent_version?: string;
}

// 历史版本列表响应
export interface PromptHistoryListResponse {
  current_version: string;
  versions: PromptHistorySummary[];
}

// ============ 提示词生成请求/响应 ============

// 继续对话生成请求
export interface ContinuePromptsRequest {
  instruction: string;
}

// 继续对话生成响应
export interface ContinuePromptsResponse {
  success: boolean;
  version: string;
  storyboard_count: number;
  message: string;
  error?: string;
}

// 重新生成请求
export interface RegeneratePromptsFromVersionRequest {
  from_version: string;
  instruction: string;
}

// 重新生成响应
export interface RegeneratePromptsFromVersionResponse {
  success: boolean;
  version: string;
  deleted_versions: string[];
  storyboard_count: number;
  message: string;
  error?: string;
}

// 切换版本请求
export interface SwitchVersionRequest {
  version: string;
}

// 切换版本响应（后端返回简单字符串，前端包装）
export type SwitchVersionResponse = string;

// ============ 项目复制类型 ============

// 复制项目请求
export interface CopyProjectRequest {
  name: string;
}

// 复制项目响应
export interface CopyProjectResponse {
  success: boolean;
  source_project_id: string;
  new_project_id: string;
  message: string;
}

// ============ 分镜管理类型 ============

// 删除分镜响应
export interface DeleteStoryboardResponse {
  success: boolean;
  deleted_index: number;
  new_storyboard_count: number;
  message: string;
  error?: string;
}

// 新增分镜请求
export interface AddStoryboardRequest {
  position: number;
  insert_type: InsertType;
}

// 新增分镜响应
export interface AddStoryboardResponse {
  success: boolean;
  new_index: number;
  new_storyboard_count: number;
  storyboard: Storyboard;
  message?: string;
  error?: string;
}

// 交换分镜请求
export interface SwapStoryboardsRequest {
  index_a: number;
  index_b: number;
}

// 交换分镜响应
export interface SwapStoryboardsResponse {
  success: boolean;
  swapped: [number, number];
  message: string;
  error?: string;
}

// ============ 主体类型常量 ============

export const SUBJECT_TYPE_LABELS: Record<SubjectType, string> = {
  character: '角色',
  object: '物品',
  scene: '场景'
};

export const ASPECT_RATIO_LABELS: Record<AspectRatio, string> = {
  '9:16': '竖屏 9:16',
  '16:9': '横屏 16:9'
};

File: api/youtube.py
"""
YouTube AI视频制作工具 - FastAPI路由模块（简化版）

设计原则：
1. 单表JSONB - 一个project_id对应一条记录
2. 同步调用 - 前端逐个分镜调用，无需任务管理
3. 8个接口 - 项目CRUD(5) + 生成(3)

接口列表：
- POST   /api/youtube/projects                      创建项目
- GET    /api/youtube/projects                      项目列表
- GET    /api/youtube/projects/{id}                 项目详情
- PUT    /api/youtube/projects/{id}                 更新项目
- DELETE /api/youtube/projects/{id}                 删除项目
- POST   /api/youtube/projects/{id}/generate/prompts  生成提示词
- POST   /api/youtube/projects/{id}/generate/image    生成单个分镜图片
- POST   /api/youtube/projects/{id}/generate/video    生成单个分镜视频
"""

import uuid
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, HTTPException, Query

from .youtube_models import (
    ProjectStatus,
    GenerationType,
    ProjectData,
    Storyboard,
    GeneratedImage,
    GeneratedVideo,
    CreateProjectRequest,
    UpdateProjectRequest,
    GeneratePromptsRequest,
    GenerateImageRequest,
    GenerateVideoRequest,
    ProjectResponse,
    PaginatedResponse,
    GeneratePromptsResponse,
    GenerateImageResponse,
    GenerateVideoResponse,
)

router = APIRouter(prefix="/api/youtube", tags=["YouTube"])

# ============ 内存存储（开发阶段，生产环境替换为数据库） ============
_projects: dict = {}


def _generate_id() -> str:
    """生成项目ID"""
    return f"proj_{uuid.uuid4().hex[:12]}"


def _calculate_status(data: dict) -> ProjectStatus:
    """根据storyboards数据计算项目状态"""
    storyboards = data.get("storyboards", [])
    
    if not storyboards:
        return ProjectStatus.CREATED
    
    all_have_images = all(len(sb.get("images", [])) > 0 for sb in storyboards)
    all_have_videos = all(len(sb.get("videos", [])) > 0 for sb in storyboards)
    any_have_images = any(len(sb.get("images", [])) > 0 for sb in storyboards)
    any_have_videos = any(len(sb.get("videos", [])) > 0 for sb in storyboards)
    
    if all_have_videos:
        return ProjectStatus.COMPLETED
    elif any_have_videos:
        return ProjectStatus.VIDEOS_PARTIAL
    elif all_have_images:
        return ProjectStatus.IMAGES_READY
    elif any_have_images:
        return ProjectStatus.IMAGES_PARTIAL
    else:
        return ProjectStatus.PROMPTS_READY


# ============ 项目管理 API ============

@router.post(
    "/projects",
    response_model=ProjectResponse,
    summary="创建项目",
    tags=["YouTube - Projects"]
)
async def create_project(request: CreateProjectRequest) -> ProjectResponse:
    """创建项目，保存youtube_url"""
    project_id = _generate_id()
    now = datetime.now()
    
    data = ProjectData(
        name=request.name,
        youtube_url=request.youtube_url,
        status=ProjectStatus.CREATED,
        prompt_version=None,
        storyboards=[]
    )
    
    _projects[project_id] = {
        "id": project_id,
        "data": data.model_dump(),
        "created_at": now,
        "updated_at": now
    }
    
    return ProjectResponse(
        id=project_id,
        data=data,
        created_at=now,
        updated_at=now
    )


@router.get(
    "/projects",
    response_model=PaginatedResponse,
    summary="项目列表",
    tags=["YouTube - Projects"]
)
async def get_projects(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
    status: Optional[ProjectStatus] = None
) -> PaginatedResponse:
    """获取项目列表（分页）"""
    projects = list(_projects.values())
    
    if status:
        projects = [p for p in projects if p["data"]["status"] == status.value]
    
    projects.sort(key=lambda x: x["created_at"], reverse=True)
    
    total = len(projects)
    start = (page - 1) * page_size
    end = start + page_size
    
    return PaginatedResponse(
        data=[ProjectResponse(**p) for p in projects[start:end]],
        total=total,
        page=page,
        page_size=page_size
    )


@router.get(
    "/projects/{project_id}",
    response_model=ProjectResponse,
    summary="项目详情",
    tags=["YouTube - Projects"]
)
async def get_project(project_id: str) -> ProjectResponse:
    """获取项目详情（含完整storyboards）"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    return ProjectResponse(**_projects[project_id])


@router.put(
    "/projects/{project_id}",
    response_model=ProjectResponse,
    summary="更新项目",
    description="部分更新：编辑提示词、选择图片/视频",
    tags=["YouTube - Projects"]
)
async def update_project(project_id: str, request: UpdateProjectRequest) -> ProjectResponse:
    """更新项目（编辑提示词、选择图片/视频）"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    
    project = _projects[project_id]
    data = project["data"]
    
    # 更新项目名称
    if request.name is not None:
        data["name"] = request.name
    
    # 更新分镜数据
    if request.storyboards is not None:
        for update_sb in request.storyboards:
            idx = update_sb.get("index")
            if idx is None or idx >= len(data["storyboards"]):
                continue
            
            sb = data["storyboards"][idx]
            
            # 更新提示词
            if "text_to_image" in update_sb:
                sb["text_to_image"] = update_sb["text_to_image"]
                sb["is_prompt_edited"] = True
            if "image_to_video" in update_sb:
                sb["image_to_video"] = update_sb["image_to_video"]
                sb["is_prompt_edited"] = True
            
            # 更新角色引用
            if "character_refs" in update_sb:
                sb["character_refs"] = update_sb["character_refs"]
            
            # 更新选中的图片/视频索引
            if "selected_image_index" in update_sb:
                sb["selected_image_index"] = update_sb["selected_image_index"]
            if "selected_video_index" in update_sb:
                sb["selected_video_index"] = update_sb["selected_video_index"]
    
    project["updated_at"] = datetime.now()
    
    return ProjectResponse(**project)


@router.delete(
    "/projects/{project_id}",
    summary="删除项目",
    tags=["YouTube - Projects"]
)
async def delete_project(project_id: str):
    """删除项目"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    
    del _projects[project_id]
    return {"message": "项目已删除"}


# ============ 生成接口 ============

async def _call_video_analysis(youtube_url: str, instruction: str = None, system_prompt: str = None) -> dict:
    """
    调用AI代理的视频分析接口 POST /api/ai/video-analysis
    
    TODO: 替换为实际的httpx调用
    """
    # 模拟LLM返回
    return {
        "success": True,
        "content": {
            "storyboards": [
                {"index": 0, "text_to_image": "一个宁静的森林场景，阳光透过树叶洒落...", "image_to_video": "镜头缓缓向前推进，树叶轻轻摇曳..."},
                {"index": 1, "text_to_image": "城市街道的夜景，霓虹灯闪烁...", "image_to_video": "行人匆匆走过，车流穿梭..."},
                {"index": 2, "text_to_image": "海边日落，金色阳光洒在海面上...", "image_to_video": "海浪轻轻拍打沙滩，海鸥飞过..."},
            ]
        }
    }


async def _call_image_generation(image_url: str, prompt: str) -> dict:
    """
    调用AI代理的图文生图接口 POST /api/ai/image-generation
    
    TODO: 替换为实际的httpx调用
    """
    return {
        "success": True,
        "image_url": f"https://placeholder.com/img_{uuid.uuid4().hex[:8]}.png"
    }


async def _call_video_generation(image_url: str, prompt: str) -> dict:
    """
    调用AI代理的图文生视频接口 POST /api/ai/video-generation
    
    TODO: 替换为实际的httpx调用
    """
    return {
        "success": True,
        "video_url": f"https://placeholder.com/vid_{uuid.uuid4().hex[:8]}.mp4"
    }


@router.post(
    "/projects/{project_id}/generate/prompts",
    response_model=GeneratePromptsResponse,
    summary="生成提示词",
    description="调用 /api/ai/video-analysis 分析视频，初始化storyboards数组",
    tags=["YouTube - Generate"]
)
async def generate_prompts(project_id: str, request: GeneratePromptsRequest) -> GeneratePromptsResponse:
    """生成提示词（同步），初始化storyboards数组"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    
    project = _projects[project_id]
    data = project["data"]
    
    # 调用AI接口
    result = await _call_video_analysis(
        youtube_url=data["youtube_url"],
        instruction=request.instruction,
        system_prompt=request.system_prompt
    )
    
    if not result.get("success"):
        raise HTTPException(status_code=500, detail=result.get("error", "AI接口调用失败"))
    
    # 初始化storyboards数组
    llm_storyboards = result["content"]["storyboards"]
    storyboards = []
    
    for item in llm_storyboards:
        storyboards.append({
            "index": item["index"],
            "text_to_image": item["text_to_image"],
            "image_to_video": item["image_to_video"],
            "character_refs": None,
            "is_prompt_edited": False,
            "images": [],                    # 初始为空数组
            "selected_image_index": None,
            "videos": [],                    # 初始为空数组
            "selected_video_index": None
        })
    
    # 更新项目数据
    data["storyboards"] = storyboards
    data["prompt_version"] = request.version
    data["status"] = ProjectStatus.PROMPTS_READY.value
    project["updated_at"] = datetime.now()
    
    return GeneratePromptsResponse(
        success=True,
        storyboard_count=len(storyboards),
        message=f"成功生成 {len(storyboards)} 个分镜提示词"
    )


@router.post(
    "/projects/{project_id}/generate/image",
    response_model=GenerateImageResponse,
    summary="生成单个分镜图片",
    description="调用 /api/ai/image-generation，追加到images数组",
    tags=["YouTube - Generate"]
)
async def generate_image(project_id: str, request: GenerateImageRequest) -> GenerateImageResponse:
    """生成单个分镜的图片（同步），追加到images数组"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    
    project = _projects[project_id]
    data = project["data"]
    storyboards = data.get("storyboards", [])
    
    # 检查分镜索引
    if request.storyboard_index >= len(storyboards):
        raise HTTPException(status_code=400, detail=f"分镜索引 {request.storyboard_index} 不存在")
    
    sb = storyboards[request.storyboard_index]
    
    # 判断生成类型
    has_character_refs = request.character_refs and len(request.character_refs) > 0
    generation_type = GenerationType.IMAGE_TEXT_TO_IMAGE if has_character_refs else GenerationType.TEXT_TO_IMAGE
    
    # 准备角色图片URL
    image_input = ""
    if has_character_refs:
        image_input = request.character_refs[0].image_url
    
    # 调用AI接口
    result = await _call_image_generation(
        image_url=image_input,
        prompt=sb["text_to_image"]
    )
    
    if not result.get("success"):
        raise HTTPException(status_code=500, detail=result.get("error", "图片生成失败"))
    
    # 追加新图片到数组
    new_image = {
        "url": result["image_url"],
        "generation_type": generation_type.value
    }
    sb["images"].append(new_image)
    
    # 如果是第一张图片，自动选中
    if len(sb["images"]) == 1:
        sb["selected_image_index"] = 0
    
    # 更新角色引用（如果有）
    if has_character_refs:
        sb["character_refs"] = [ref.identifier for ref in request.character_refs]
    
    # 更新项目状态
    data["status"] = _calculate_status(data).value
    project["updated_at"] = datetime.now()
    
    return GenerateImageResponse(
        success=True,
        storyboard_index=request.storyboard_index,
        image=GeneratedImage(**new_image),
        message=f"分镜 {request.storyboard_index} 图片生成成功"
    )


@router.post(
    "/projects/{project_id}/generate/video",
    response_model=GenerateVideoResponse,
    summary="生成单个分镜视频",
    description="调用 /api/ai/video-generation，追加到videos数组",
    tags=["YouTube - Generate"]
)
async def generate_video(project_id: str, request: GenerateVideoRequest) -> GenerateVideoResponse:
    """生成单个分镜的视频（同步），追加到videos数组"""
    if project_id not in _projects:
        raise HTTPException(status_code=404, detail="项目不存在")
    
    project = _projects[project_id]
    data = project["data"]
    storyboards = data.get("storyboards", [])
    
    # 检查分镜索引
    if request.storyboard_index >= len(storyboards):
        raise HTTPException(status_code=400, detail=f"分镜索引 {request.storyboard_index} 不存在")
    
    sb = storyboards[request.storyboard_index]
    
    # 检查是否有选中的图片
    if sb["selected_image_index"] is None or len(sb["images"]) == 0:
        raise HTTPException(status_code=400, detail=f"分镜 {request.storyboard_index} 没有选中的图片")
    
    # 获取选中的图片URL
    selected_image = sb["images"][sb["selected_image_index"]]
    source_image_url = selected_image["url"]
    
    # 调用AI接口
    result = await _call_video_generation(
        image_url=source_image_url,
        prompt=sb["image_to_video"]
    )
    
    if not result.get("success"):
        raise HTTPException(status_code=500, detail=result.get("error", "视频生成失败"))
    
    # 追加新视频到数组
    new_video = {
        "url": result["video_url"],
        "source_image_index": sb["selected_image_index"]
    }
    sb["videos"].append(new_video)
    
    # 如果是第一个视频，自动选中
    if len(sb["videos"]) == 1:
        sb["selected_video_index"] = 0
    
    # 更新项目状态
    data["status"] = _calculate_status(data).value
    project["updated_at"] = datetime.now()
    
    return GenerateVideoResponse(
        success=True,
        storyboard_index=request.storyboard_index,
        video=GeneratedVideo(**new_video),
        message=f"分镜 {request.storyboard_index} 视频生成成功"
    )
